File: src/lib/e2ee/v2/pairing-qr.ts
'use client';

/**
 * QR pairing (device → device) для E2EE v2.
 *
 * Сценарий: пользователь хочет перенести приватник со старого устройства (donor)
 * на новое (initiator). Новое устройство генерирует эфемерный ECDH-ключ и
 * создаёт `users/{uid}/e2eePairingSessions/{sessionId}`; старое сканирует QR,
 * выполняет ECDH со своим одноразовым ключом, шифрует приватник под общий AES
 * и записывает донорскую часть обратно в документ.
 *
 * Почему не peer-to-peer:
 *  - Firestore уже обеспечивает authenticated RPC (правила — только владелец);
 *  - ни web-камера, ни мобильная камера не могут стабильно сделать прямой p2p;
 *  - TTL-очистка документов делается Cloud Function (см.
 *    `functions/src/triggers/scheduled/cleanupE2eePairingSessions.ts`), чтобы
 *    мусор не копился.
 *
 * Защита от MITM:
 *  - 6-значный код — это HMAC-SHA-256 от общего AES-ключа, обрезанный до 6
 *    десятичных цифр. Пользователь сверяет его на обоих экранах. Если код
 *    совпал — ECDH-handshake пришёл от того устройства, которое сейчас в руках
 *    у пользователя, а не от проникшего посередине.
 *
 * Сам модуль не знает ничего про UI (QR-рендер, сканер камеры) — только
 * протокол и Firestore. UI-слой поверх сможет упаковать `buildQrPayload`
 * в QR и распарсить `parseQrPayload` обратно.
 */

import {
  collection,
  doc,
  deleteDoc,
  getDoc,
  onSnapshot,
  setDoc,
  type Firestore,
  type Unsubscribe,
} from 'firebase/firestore';
import type { E2eePairingSessionDocV2 } from '@/lib/types';
import { fromBase64, toBase64 } from '@/lib/e2ee/b64';
import { exportSpkiPublic, generateEcdhKeyPair, importEcdhPublicFromSpki } from '@/lib/e2ee/webcrypto';

/** Версия протокола pairing — дублируется внутрь QR-payload для forward-compat. */
export const E2EE_PAIRING_QR_VERSION = 'v2-pairing-1' as const;

/** Время жизни сессии на клиенте — сервер (Cloud Function) чистит по этому же сроку. */
export const E2EE_PAIRING_TTL_MS = 10 * 60 * 1000;

/** Публичный payload, который помещается в QR. Размер ≈ 170 байт base64-url. */
export type PairingQrPayload = {
  v: typeof E2EE_PAIRING_QR_VERSION;
  uid: string;
  sessionId: string;
  initiatorEphPub: string;
};

function randomSessionId(): string {
  // 12 байт — короче ULID, но достаточно, чтобы не столкнуться. Firestore doc id.
  const bytes = new Uint8Array(12);
  crypto.getRandomValues(bytes);
  return Array.from(bytes, (b) => b.toString(16).padStart(2, '0')).join('');
}

/** Сериализация QR-payload в компактную base64-url строку. */
export function buildQrPayload(payload: PairingQrPayload): string {
  return toBase64(new TextEncoder().encode(JSON.stringify(payload)))
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/, '');
}

/** Разбор строки, полученной со сканера. Throws при невалидном формате. */
export function parseQrPayload(raw: string): PairingQrPayload {
  const normalized = raw.replace(/-/g, '+').replace(/_/g, '/');
  const padded = normalized + '='.repeat((4 - (normalized.length % 4)) % 4);
  const decoded = new TextDecoder().decode(fromBase64(padded));
  const parsed = JSON.parse(decoded) as Partial<PairingQrPayload>;
  if (
    parsed.v !== E2EE_PAIRING_QR_VERSION ||
    typeof parsed.uid !== 'string' ||
    typeof parsed.sessionId !== 'string' ||
    typeof parsed.initiatorEphPub !== 'string'
  ) {
    throw new Error('E2EE_PAIRING_INVALID_QR');
  }
  return parsed as PairingQrPayload;
}

/**
 * ECDH P-256 → 32 байта общего ключа. Используется одинаково для derive на donor
 * и initiator стороне. Делаем импорт/деривацию внутри, чтобы вызывающий не
 * таскал CryptoKey.
 */
async function deriveSharedAesRaw(
  myPrivate: CryptoKey,
  peerPublicSpki: Uint8Array
): Promise<Uint8Array> {
  const peer = await importEcdhPublicFromSpki(peerPublicSpki);
  const bits = await crypto.subtle.deriveBits(
    { name: 'ECDH', public: peer } as EcdhKeyDeriveParams,
    myPrivate,
    256
  );
  return new Uint8Array(bits);
}

/** 6-значный код для проверки пары глазами (HMAC(key, "pairing-code") mod 10^6). */
async function shortPairingCode(sharedKey: Uint8Array): Promise<string> {
  const hmacKey = await crypto.subtle.importKey(
    'raw',
    sharedKey.buffer.slice(sharedKey.byteOffset, sharedKey.byteOffset + sharedKey.byteLength) as ArrayBuffer,
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign']
  );
  const tag = new Uint8Array(
    await crypto.subtle.sign(
      'HMAC',
      hmacKey,
      new TextEncoder().encode('lighchat/v2/pairing-code') as BufferSource
    )
  );
  // первые 4 байта → uint32 → mod 1_000_000 → 6 цифр
  const n = ((tag[0] << 24) | (tag[1] << 16) | (tag[2] << 8) | tag[3]) >>> 0;
  return (n % 1_000_000).toString().padStart(6, '0');
}

/* -------------------------------------------------------------------------- */
/*                               INITIATOR (new device)                       */
/* -------------------------------------------------------------------------- */

export type InitiatorSession = {
  sessionId: string;
  qrPayload: PairingQrPayload;
  qrEncoded: string;
  /** Эфемерный ECDH-keypair — держим в памяти до завершения или таймаута. */
  ephemeralPrivate: CryptoKey;
  ephemeralPublicSpkiB64: string;
};

/**
 * Вызывается на новом устройстве. Генерирует эфемерный ключ, создаёт документ
 * `pairingSessions/{sessionId}` в состоянии `awaiting_scan`, возвращает QR payload.
 */
export async function initiatePairingSessionV2(
  firestore: Firestore,
  userId: string
): Promise<InitiatorSession> {
  const keyPair = await generateEcdhKeyPair();
  const pubBytes = await exportSpkiPublic(keyPair.publicKey);
  const pubB64 = toBase64(pubBytes);
  const sessionId = randomSessionId();
  const nowIso = new Date().toISOString();
  const expiresIso = new Date(Date.now() + E2EE_PAIRING_TTL_MS).toISOString();
  const doc0: E2eePairingSessionDocV2 = {
    sessionId,
    createdAt: nowIso,
    expiresAt: expiresIso,
    state: 'awaiting_scan',
    initiatorEphPubSpkiB64: pubB64,
  };
  await setDoc(
    doc(firestore, 'users', userId, 'e2eePairingSessions', sessionId),
    doc0
  );
  const qrPayload: PairingQrPayload = {
    v: E2EE_PAIRING_QR_VERSION,
    uid: userId,
    sessionId,
    initiatorEphPub: pubB64,
  };
  return {
    sessionId,
    qrPayload,
    qrEncoded: buildQrPayload(qrPayload),
    ephemeralPrivate: keyPair.privateKey,
    ephemeralPublicSpkiB64: pubB64,
  };
}

/**
 * Подписка на обновления pairing-документа. Возвращает unsubscribe.
 * UI использует, чтобы показать "старое устройство отсканировало QR".
 */
export function watchPairingSessionV2(
  firestore: Firestore,
  userId: string,
  sessionId: string,
  onUpdate: (data: E2eePairingSessionDocV2 | null) => void
): Unsubscribe {
  return onSnapshot(
    doc(firestore, 'users', userId, 'e2eePairingSessions', sessionId),
    (snap) => {
      if (!snap.exists()) {
        onUpdate(null);
        return;
      }
      onUpdate(snap.data() as E2eePairingSessionDocV2);
    }
  );
}

/**
 * Вызывается на initiator после того, как donor прислал свою часть. Выполняет
 * финальный ECDH, расшифровывает payload от donor'а.
 * Возвращает приватник и 6-значный код для сверки пользователем.
 */
export async function consumeDonorPayloadV2(params: {
  firestore: Firestore;
  userId: string;
  sessionId: string;
  initiatorEphemeralPrivate: CryptoKey;
  donorDoc: E2eePairingSessionDocV2;
}): Promise<{ privateKeyPkcs8: Uint8Array; pairingCode: string }> {
  const donor = params.donorDoc.donorPayload;
  if (!donor) {
    throw new Error('E2EE_PAIRING_DONOR_PAYLOAD_MISSING');
  }
  const donorEphPubSpki = fromBase64(donor.donorEphPubSpkiB64);
  const sharedRaw = await deriveSharedAesRaw(params.initiatorEphemeralPrivate, donorEphPubSpki);
  const aesKey = await crypto.subtle.importKey(
    'raw',
    sharedRaw.buffer.slice(
      sharedRaw.byteOffset,
      sharedRaw.byteOffset + sharedRaw.byteLength
    ) as ArrayBuffer,
    { name: 'AES-GCM', length: 256 },
    false,
    ['decrypt']
  );
  const iv = fromBase64(donor.ivB64);
  const ct = fromBase64(donor.ciphertextB64);
  const aad = new TextEncoder().encode(
    `lighchat/v2/pairing|${params.userId}|${params.sessionId}`
  );
  let privateKeyPkcs8: Uint8Array;
  try {
    const pt = await crypto.subtle.decrypt(
      {
        name: 'AES-GCM',
        iv: iv.buffer.slice(iv.byteOffset, iv.byteOffset + iv.byteLength) as ArrayBuffer,
        additionalData: aad.buffer.slice(
          aad.byteOffset,
          aad.byteOffset + aad.byteLength
        ) as ArrayBuffer,
      },
      aesKey,
      ct.buffer.slice(ct.byteOffset, ct.byteOffset + ct.byteLength) as ArrayBuffer
    );
    privateKeyPkcs8 = new Uint8Array(pt);
  } catch {
    throw new Error('E2EE_PAIRING_DECRYPT_FAILED');
  }
  const pairingCode = await shortPairingCode(sharedRaw);

  // Завершаем сессию: меняем state, чтобы donor видел финал. Удалять доверяем
  // Cloud Function / TTL.
  await setDoc(
    doc(params.firestore, 'users', params.userId, 'e2eePairingSessions', params.sessionId),
    { ...params.donorDoc, state: 'completed' },
    { merge: true }
  );
  return { privateKeyPkcs8, pairingCode };
}

/* -------------------------------------------------------------------------- */
/*                                   DONOR                                    */
/* -------------------------------------------------------------------------- */

export type DonorRespondOptions = {
  firestore: Firestore;
  userId: string;
  sessionId: string;
  initiatorEphPubSpkiB64: string;
  privateKeyPkcs8: Uint8Array;
  deviceDraft: E2eePairingSessionDocV2['donorPayload'] extends infer T
    ? T extends { deviceDraft: infer D }
      ? D
      : never
    : never;
};

/**
 * Вызывается на donor-устройстве после сканирования QR. Генерирует эфемерный
 * ключ donor'a, шифрует приватник под общий AES, пишет обратно в документ.
 * Возвращает 6-значный код для показа пользователю на сверку.
 */
export async function donorRespondToPairingV2(
  opts: DonorRespondOptions
): Promise<{ pairingCode: string }> {
  const donorKeyPair = await generateEcdhKeyPair();
  const donorPubBytes = await exportSpkiPublic(donorKeyPair.publicKey);
  const initiatorPub = fromBase64(opts.initiatorEphPubSpkiB64);
  const sharedRaw = await deriveSharedAesRaw(donorKeyPair.privateKey, initiatorPub);
  const aesKey = await crypto.subtle.importKey(
    'raw',
    sharedRaw.buffer.slice(
      sharedRaw.byteOffset,
      sharedRaw.byteOffset + sharedRaw.byteLength
    ) as ArrayBuffer,
    { name: 'AES-GCM', length: 256 },
    false,
    ['encrypt']
  );
  const iv = new Uint8Array(12);
  crypto.getRandomValues(iv);
  const aad = new TextEncoder().encode(
    `lighchat/v2/pairing|${opts.userId}|${opts.sessionId}`
  );
  const ct = new Uint8Array(
    await crypto.subtle.encrypt(
      {
        name: 'AES-GCM',
        iv: iv.buffer.slice(iv.byteOffset, iv.byteOffset + iv.byteLength) as ArrayBuffer,
        additionalData: aad.buffer.slice(
          aad.byteOffset,
          aad.byteOffset + aad.byteLength
        ) as ArrayBuffer,
      },
      aesKey,
      opts.privateKeyPkcs8.buffer.slice(
        opts.privateKeyPkcs8.byteOffset,
        opts.privateKeyPkcs8.byteOffset + opts.privateKeyPkcs8.byteLength
      ) as ArrayBuffer
    )
  );

  const ref = doc(
    opts.firestore,
    'users',
    opts.userId,
    'e2eePairingSessions',
    opts.sessionId
  );
  const snap = await getDoc(ref);
  if (!snap.exists()) {
    throw new Error('E2EE_PAIRING_SESSION_NOT_FOUND');
  }
  const existing = snap.data() as E2eePairingSessionDocV2;
  const donorPayload: E2eePairingSessionDocV2['donorPayload'] = {
    donorEphPubSpkiB64: toBase64(donorPubBytes),
    ivB64: toBase64(iv),
    ciphertextB64: toBase64(ct),
    deviceDraft: opts.deviceDraft,
  };
  await setDoc(
    ref,
    <E2eePairingSessionDocV2>{
      ...existing,
      state: 'awaiting_accept',
      donorPayload,
    },
    { merge: true }
  );
  const pairingCode = await shortPairingCode(sharedRaw);
  return { pairingCode };
}

/**
 * Ручная отмена сессии (пользователь нажал "Отмена" или код не совпал).
 * Помечаем `rejected` и удаляем сразу — TTL не нужен.
 */
export async function rejectPairingSessionV2(
  firestore: Firestore,
  userId: string,
  sessionId: string
): Promise<void> {
  const ref = doc(firestore, 'users', userId, 'e2eePairingSessions', sessionId);
  try {
    const snap = await getDoc(ref);
    if (!snap.exists()) return;
    await setDoc(ref, { ...(snap.data() as E2eePairingSessionDocV2), state: 'rejected' }, { merge: true });
  } finally {
    await deleteDoc(ref).catch(() => {
      /* idempotent */
    });
  }
}

/** Служебный accessor, удобен для unit-тестов. */
export async function listActivePairingSessionsV2(
  firestore: Firestore,
  userId: string
): Promise<E2eePairingSessionDocV2[]> {
  const snap = await import('firebase/firestore').then(({ getDocs }) =>
    getDocs(collection(firestore, 'users', userId, 'e2eePairingSessions'))
  );
  return snap.docs.map((d) => d.data() as E2eePairingSessionDocV2);
}

File: mobile/packages/lighchat_firebase/lib/src/e2ee/pairing_qr.dart
/// QR pairing (device → device) для E2EE v2 — mobile.
///
/// Зеркалит `src/lib/e2ee/v2/pairing-qr.ts`. Совместимость с веб-клиентом
/// гарантируется одинаковым форматом `users/{uid}/e2eePairingSessions/{id}`,
/// одинаковым AAD (`lighchat/v2/pairing|{uid}|{sessionId}`) и одинаковой
/// сериализацией QR-payload.
///
/// Этот модуль реализует чистый **протокол**: эфемерный ECDH, шифрование
/// приватника, 6-значный код для сверки. Камера/QR-рендер живут на UI-слое
/// (`mobile/app/lib/features/settings/ui/devices_screen.dart` сможет
/// добавить сканирование через `mobile_scanner` позже, без изменений здесь).
library;

import 'dart:convert';
import 'dart:typed_data';

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/foundation.dart' show debugPrint;
import 'package:pointycastle/export.dart';

import 'webcrypto_compat.dart';

/// Версия протокола pairing — дублируется внутрь QR payload, чтобы старые
/// клиенты могли отвергнуть неизвестный формат.
const String e2eePairingQrVersion = 'v2-pairing-1';

/// TTL сессии. Сервер / cleanup-CF чистит по этому же сроку.
const Duration e2eePairingTtl = Duration(minutes: 10);

/// Публичный payload, который кодируется в QR. Идентичен web-формату.
class PairingQrPayload {
  const PairingQrPayload({
    required this.uid,
    required this.sessionId,
    required this.initiatorEphPubSpkiB64,
  });

  final String uid;
  final String sessionId;
  final String initiatorEphPubSpkiB64;

  Map<String, Object?> toJson() => <String, Object?>{
        'v': e2eePairingQrVersion,
        'uid': uid,
        'sessionId': sessionId,
        'initiatorEphPub': initiatorEphPubSpkiB64,
      };

  static PairingQrPayload fromJson(Map<String, Object?> json) {
    final version = json['v'];
    if (version != e2eePairingQrVersion) {
      throw StateError('E2EE_PAIRING_INVALID_QR');
    }
    final uid = json['uid'];
    final sessionId = json['sessionId'];
    final initiatorEphPub = json['initiatorEphPub'];
    if (uid is! String || sessionId is! String || initiatorEphPub is! String) {
      throw StateError('E2EE_PAIRING_INVALID_QR');
    }
    return PairingQrPayload(
      uid: uid,
      sessionId: sessionId,
      initiatorEphPubSpkiB64: initiatorEphPub,
    );
  }
}

/// base64-url сериализация (compatible с web `buildQrPayload`).
String buildQrPayload(PairingQrPayload payload) {
  final json = jsonEncode(payload.toJson());
  final b64 = base64.encode(utf8.encode(json));
  return b64.replaceAll('+', '-').replaceAll('/', '_').replaceAll('=', '');
}

/// Разбор строки со сканера. Throws `E2EE_PAIRING_INVALID_QR` на битом входе.
PairingQrPayload parseQrPayload(String raw) {
  final normalized = raw.replaceAll('-', '+').replaceAll('_', '/');
  final padLen = (4 - normalized.length % 4) % 4;
  final padded = normalized + '=' * padLen;
  try {
    final decoded = utf8.decode(base64.decode(padded));
    final obj = jsonDecode(decoded);
    if (obj is! Map) {
      throw StateError('E2EE_PAIRING_INVALID_QR');
    }
    return PairingQrPayload.fromJson(Map<String, Object?>.from(obj));
  } catch (_) {
    throw StateError('E2EE_PAIRING_INVALID_QR');
  }
}

String _randomSessionId() {
  final bytes = randomBytes(12);
  final sb = StringBuffer();
  for (final b in bytes) {
    sb.write(b.toRadixString(16).padLeft(2, '0'));
  }
  return sb.toString();
}

Uint8List _pairingAad(String userId, String sessionId) {
  return Uint8List.fromList(
    utf8.encode('lighchat/v2/pairing|$userId|$sessionId'),
  );
}

/// Короткий 6-значный код, сверяется глазами. HMAC(sharedKey, "pairing-code"),
/// первые 4 байта → mod 10^6 → 6 цифр. Тот же алгоритм, что в web.
Future<String> _shortPairingCode(Uint8List sharedKey) async {
  final hmac = HMac(SHA256Digest(), 64)..init(KeyParameter(sharedKey));
  final msg = Uint8List.fromList(utf8.encode('lighchat/v2/pairing-code'));
  hmac.update(msg, 0, msg.length);
  final tag = Uint8List(hmac.macSize);
  hmac.doFinal(tag, 0);
  final n = (tag[0] << 24 | tag[1] << 16 | tag[2] << 8 | tag[3]) & 0xFFFFFFFF;
  return (n % 1000000).toString().padLeft(6, '0');
}

/* -------------------------------------------------------------------------- */
/*                              INITIATOR (new device)                        */
/* -------------------------------------------------------------------------- */

/// Состояние, возвращаемое `initiateMobilePairingSession`. UI хранит его в
/// памяти до истечения TTL или ручной отмены.
class MobileInitiatorSession {
  MobileInitiatorSession({
    required this.sessionId,
    required this.payload,
    required this.encoded,
    required this.ephemeralKeyPair,
  });

  final String sessionId;
  final PairingQrPayload payload;
  final String encoded;
  final EcdhP256KeyPair ephemeralKeyPair;
}

/// Вызывается на новом устройстве. Создаёт документ сессии и возвращает
/// данные для отрисовки QR.
Future<MobileInitiatorSession> initiateMobilePairingSession({
  required FirebaseFirestore firestore,
  required String userId,
}) async {
  debugPrint('[QR-PAIR/init] step=1 generate ECDH P-256 keypair');
  final EcdhP256KeyPair eph;
  try {
    eph = await generateEcdhP256KeyPair();
  } catch (e, st) {
    debugPrint('[QR-PAIR/init] FAIL generateEcdhP256KeyPair: $e\n$st');
    rethrow;
  }
  debugPrint('[QR-PAIR/init] step=2 export SPKI public');
  final Uint8List pubSpki;
  try {
    pubSpki = await eph.exportSpkiPublic();
  } catch (e, st) {
    debugPrint('[QR-PAIR/init] FAIL exportSpkiPublic: $e\n$st');
    rethrow;
  }
  final pubB64 = base64.encode(pubSpki);
  final sessionId = _randomSessionId();
  final nowIso = DateTime.now().toUtc().toIso8601String();
  final expiresIso = DateTime.now()
      .toUtc()
      .add(e2eePairingTtl)
      .toIso8601String();
  final path = 'users/$userId/e2eePairingSessions/$sessionId';
  debugPrint('[QR-PAIR/init] step=3 firestore.set path=$path pubB64.len=${pubB64.length}');
  try {
    await firestore
        .collection('users')
        .doc(userId)
        .collection('e2eePairingSessions')
        .doc(sessionId)
        .set(<String, Object?>{
      'sessionId': sessionId,
      'createdAt': nowIso,
      'expiresAt': expiresIso,
      'state': 'awaiting_scan',
      'initiatorEphPubSpkiB64': pubB64,
    });
  } catch (e, st) {
    debugPrint('[QR-PAIR/init] FAIL firestore.set path=$path: $e\n$st');
    rethrow;
  }
  debugPrint('[QR-PAIR/init] step=4 firestore.set OK; building QR payload');
  final payload = PairingQrPayload(
    uid: userId,
    sessionId: sessionId,
    initiatorEphPubSpkiB64: pubB64,
  );
  return MobileInitiatorSession(
    sessionId: sessionId,
    payload: payload,
    encoded: buildQrPayload(payload),
    ephemeralKeyPair: eph,
  );
}

/// Подписка на Firestore-сессию. UI смотрит на смену `state` для перехода
/// к экрану сверки кода.
Stream<Map<String, Object?>?> watchMobilePairingSession({
  required FirebaseFirestore firestore,
  required String userId,
  required String sessionId,
}) {
  return firestore
      .collection('users')
      .doc(userId)
      .collection('e2eePairingSessions')
      .doc(sessionId)
      .snapshots()
      .map((snap) => snap.data());
}

class ConsumeDonorResult {
  const ConsumeDonorResult({required this.privateKeyPkcs8, required this.pairingCode});
  final Uint8List privateKeyPkcs8;
  final String pairingCode;
}

/// Вызывается на initiator после того, как donor записал `donorPayload`.
/// Разбирает документ и возвращает расшифрованный PKCS#8 приватник.
Future<ConsumeDonorResult> consumeDonorPayloadMobile({
  required FirebaseFirestore firestore,
  required String userId,
  required String sessionId,
  required EcdhP256KeyPair initiatorEphemeral,
  required Map<String, Object?> donorDocument,
}) async {
  debugPrint('[QR-PAIR/consume] step=1 parse donorPayload');
  final donor = donorDocument['donorPayload'];
  if (donor is! Map) {
    debugPrint('[QR-PAIR/consume] FAIL donorPayload is not Map: ${donor.runtimeType}');
    throw StateError('E2EE_PAIRING_DONOR_PAYLOAD_MISSING');
  }
  final donorMap = Map<String, Object?>.from(donor);
  debugPrint('[QR-PAIR/consume] step=2 ECDH derive shared key');
  late final Uint8List pt;
  try {
    final donorEphSpki = Uint8List.fromList(
      base64.decode(donorMap['donorEphPubSpkiB64'] as String),
    );
    final donorEph = await importSpkiP256(spki: donorEphSpki);
    final shared = await ecdhP256DeriveBits32(
      privateKey: initiatorEphemeral.privateKey,
      remotePublic: donorEph,
    );
    final iv = Uint8List.fromList(base64.decode(donorMap['ivB64'] as String));
    final ct = Uint8List.fromList(base64.decode(donorMap['ciphertextB64'] as String));
    final aad = _pairingAad(userId, sessionId);
    debugPrint('[QR-PAIR/consume] step=3 AES-GCM decrypt ct.len=${ct.length} iv.len=${iv.length}');
    try {
      pt = await aesGcmDecryptV2(
        key: shared,
        iv: iv,
        ciphertextPlusTag: ct,
        aad: aad,
      );
    } catch (e, st) {
      debugPrint('[QR-PAIR/consume] FAIL aesGcmDecryptV2: $e\n$st');
      throw StateError('E2EE_PAIRING_DECRYPT_FAILED');
    }
    final code = await _shortPairingCode(shared);
    debugPrint('[QR-PAIR/consume] step=4 firestore.set state=completed sessionId=$sessionId');
    try {
      await firestore
          .collection('users')
          .doc(userId)
          .collection('e2eePairingSessions')
          .doc(sessionId)
          .set(<String, Object?>{'state': 'completed'}, SetOptions(merge: true));
    } catch (e, st) {
      debugPrint('[QR-PAIR/consume] FAIL firestore.set completed: $e\n$st');
      rethrow;
    }
    debugPrint('[QR-PAIR/consume] OK pairingCode=$code pt.len=${pt.length}');
    return ConsumeDonorResult(privateKeyPkcs8: pt, pairingCode: code);
  } catch (e, st) {
    debugPrint('[QR-PAIR/consume] outer FAIL: $e\n$st');
    rethrow;
  }
}

/* -------------------------------------------------------------------------- */
/*                                  DONOR                                     */
/* -------------------------------------------------------------------------- */

/// Мета-инфа нового устройства, которую donor-клиент встраивает в документ
/// (чтобы UI initiator'а мог отрисовать «подтверждаете добавление X?»).
class MobileDeviceDraft {
  const MobileDeviceDraft({
    required this.deviceId,
    required this.platform,
    required this.label,
    required this.publicKeySpkiB64,
  });

  final String deviceId;
  final String platform;
  final String label;
  final String publicKeySpkiB64;

  Map<String, Object?> toJson() => <String, Object?>{
        'deviceId': deviceId,
        'platform': platform,
        'label': label,
        'publicKeySpki': publicKeySpkiB64,
      };
}

/// Вызывается на donor-устройстве после сканирования QR нового устройства.
/// Генерирует эфемерный ключ, шифрует приватник, пишет `donorPayload`.
/// Возвращает 6-значный код, который надо показать для сверки.
Future<String> donorRespondToPairingMobile({
  required FirebaseFirestore firestore,
  required String userId,
  required String sessionId,
  required String initiatorEphPubSpkiB64,
  required Uint8List privateKeyPkcs8,
  required MobileDeviceDraft deviceDraft,
}) async {
  debugPrint('[QR-PAIR/donor] step=1 generate ECDH P-256 keypair');
  final donorEph = await generateEcdhP256KeyPair();
  debugPrint('[QR-PAIR/donor] step=2 export SPKI public + derive shared');
  final donorPubSpki = await donorEph.exportSpkiPublic();
  final initiatorSpki = Uint8List.fromList(
    base64.decode(initiatorEphPubSpkiB64),
  );
  final initiatorPub = await importSpkiP256(spki: initiatorSpki);
  final shared = await ecdhP256DeriveBits32(
    privateKey: donorEph.privateKey,
    remotePublic: initiatorPub,
  );
  final iv = randomBytes(12);
  final aad = _pairingAad(userId, sessionId);
  debugPrint('[QR-PAIR/donor] step=3 AES-GCM encrypt pkcs8.len=${privateKeyPkcs8.length}');
  final ct = await aesGcmEncryptV2(
    key: shared,
    iv: iv,
    plaintext: privateKeyPkcs8,
    aad: aad,
  );
  final path = 'users/$userId/e2eePairingSessions/$sessionId';
  debugPrint('[QR-PAIR/donor] step=4 firestore.set path=$path (awaiting_accept + donorPayload) ct.len=${ct.length}');
  try {
    await firestore
        .collection('users')
        .doc(userId)
        .collection('e2eePairingSessions')
        .doc(sessionId)
        .set(<String, Object?>{
      'state': 'awaiting_accept',
      'donorPayload': <String, Object?>{
        'donorEphPubSpkiB64': base64.encode(donorPubSpki),
        'ivB64': base64.encode(iv),
        'ciphertextB64': base64.encode(ct),
        'deviceDraft': deviceDraft.toJson(),
      },
    }, SetOptions(merge: true));
  } catch (e, st) {
    debugPrint('[QR-PAIR/donor] FAIL firestore.set awaiting_accept path=$path: $e\n$st');
    rethrow;
  }
  debugPrint('[QR-PAIR/donor] OK firestore.set done');
  return _shortPairingCode(shared);
}

/// Ручная отмена сессии (пользователь нажал "Отмена" или код не совпал).
/// Помечаем `rejected` и пытаемся удалить — ошибки игнорируются
/// (идемпотентно).
Future<void> rejectMobilePairingSession({
  required FirebaseFirestore firestore,
  required String userId,
  required String sessionId,
}) async {
  final ref = firestore
      .collection('users')
      .doc(userId)
      .collection('e2eePairingSessions')
      .doc(sessionId);
  try {
    await ref.set(<String, Object?>{'state': 'rejected'}, SetOptions(merge: true));
  } catch (_) {
    // ignore — документа может не быть.
  }
  try {
    await ref.delete();
  } catch (_) {
    // idempotent.
  }
}
